### PR TITLE
Added support for Rich Repr protocol

### PR DESCRIPTION
**Pull request by willmcgugan**
## Change Summary

Adds support for the [Rich Repr Protocol](https://rich.readthedocs.io/en/latest/pretty.html#rich-repr-protocol) so that Rich can pretty print PyDantic objects.

Screen Shot 2021-08-15 at 14 13 35

Note that Rich does not need to be added a dependency. If PyDantic objects are printed with Rich you will get the formatting above, otherwise there are no changes.

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by willmcgugan** [comment]
please review

Not sure if this change warrants any update to the docs. If anything, maybe a sentence or two max, but where?

**Comment by samuelcolvin** [comment]
Thanks so much.

When life is less manic I'll review and merge.

File: changes/3099-willmcgugan.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Adds a __rich_repr__ method to Representation class which enables pretty printing with [Rich](https://github.com/willmcgugan/rich)
```
Comment (PrettyWood) [inline changes/3099-willmcgugan.md:1]:
```suggestion
Adds a `__rich_repr__` method to `Representation` class which enables pretty printing with [Rich](https://github.com/willmcgugan/rich)
```

**Review by PrettyWood — COMMENTED** [review]
You could probably add a note under `Usage with devtools` (e.g. `Usage with rich`) with just a screenshot and a link to `rich`. WDYT?

**Comment by PrettyWood** [comment]
Please update

**Comment by willmcgugan** [comment]
@PrettyWood Added a Usage with rich page as suggested. Here's what it looks like:

Screen Shot 2021-09-07 at 15 19 04

**Review by PrettyWood — APPROVED** [review]

**Comment by PrettyWood** [comment]
Looks great! Thank you 🙏 (big fan of your work btw)

**Comment by manzt** [comment]
Love rich and pydantic, hope this gets merged!

**Comment by soumendra** [comment]
Great work. Hopefully we'll see this in the next pydantic release?

**Comment by samuelcolvin** [comment]
Thanks so much @willmcgugan, much appreciated.

Sorry this has taken so long to get to.

**Comment by willmcgugan** [comment]
Thanks @samuelcolvin !

**Pull request merged**